test/Events.cpp

```diff
@@ -0,0 +1,121 @@
+#include "infinispan/hotrod/ConfigurationBuilder.h"
+#include "infinispan/hotrod/RemoteCacheManager.h"
+#include "infinispan/hotrod/RemoteCache.h"
+#include "infinispan/hotrod/Version.h"
+
+#include "infinispan/hotrod/JBasicMarshaller.h"
+#include "infinispan/hotrod/CacheClientListener.h"
+#include <stdlib.h>
+#include <iostream>
+#include <memory>
+#include <typeinfo>
+#include <thread>
+#include <future>
+#include <chrono>
+#include <functional>
+
+// For CTest: return 0 if all tests pass, non-zero otherwise.
+
+#define ASSERT(VALUE, COND, RES, MSG) \
+	do { if (!(COND)){ \
+		std::cout << "FAIL :(" << #COND << "):(" << VALUE << "):" << #MSG << std::endl; \
+		RES=-1; }\
+		} while(0)
+
+using namespace infinispan::hotrod;
+
+int main(int argc, char** argv) {
+
+    int result=0;
+    std::cout << "Events and Listeners" << std::endl;
+    {
+        ConfigurationBuilder builder;
+                builder.balancingStrategyProducer(nullptr);
+        builder.addServer().host(argc > 1 ? argv[1] : "127.0.0.1").port(argc > 2 ? atoi(argv[2]) : 11222);
+        builder.protocolVersion(Configuration::PROTOCOL_VERSION_24);
+        RemoteCacheManager cacheManager(builder.build(), false);
+        cacheManager.start();
+        JBasicMarshaller<int> *km = new JBasicMarshaller<int>();
+        JBasicMarshaller<std::string> *vm = new JBasicMarshaller<std::string>();
+        RemoteCache<int, std::string> cache = cacheManager.getCache<int, std::string>(km,
+                &Marshaller<int>::destroy,
+                vm,
+                &Marshaller<std::string>::destroy);
+        cache.clear();
+        std::vector<std::vector<char> > filterFactoryParams;
+        std::vector<std::vector<char> > converterFactoryParams;
+        CacheClientListener<int, std::string> cl(cache);
+        int createdCount=0, modifiedCount=0, removedCount=0;
+        std::function<void(ClientCacheEntryCreatedEvent<int>)> listenerCreated = [&createdCount](ClientCacheEntryCreatedEvent<int> e) { createdCount++; };
+        std::function<void(ClientCacheEntryModifiedEvent<int>)> listenerModified = [&modifiedCount](ClientCacheEntryModifiedEvent <int> e) { modifiedCount++; };
+        std::function<void(ClientCacheEntryRemovedEvent<int>)> listenerRemoved = [&removedCount](ClientCacheEntryRemovedEvent <int> e) { removedCount++; };
+
+        cl.add_listener(listenerCreated);
+        cl.add_listener(listenerModified);
+        cl.add_listener(listenerRemoved);
+
+        cache.put(1,"v1");
+        cache.put(2,"v2");
+        cache.put(3,"v3");
+
+        cache.addClientListener(cl, filterFactoryParams, converterFactoryParams);
+
+        cache.put(4,"v4");
+        cache.put(5,"v5");
+        cache.put(6,"v6");
+
+        // Give some time to listener thread to update the counters
+        std::this_thread::sleep_for(std::chrono::milliseconds(100)); 
```

Comment:
Instead of waiting a fixed amount of time, maybe it's better to have a way to loop with small pauses. Maybe you can come up with an ASSERT which takes a pause time and a number of iterations before giving up.